synapse/rest/client/transactions.py

```diff
@@ -92,6 +92,14 @@ def fetch_or_execute(self, txn_key, fn, *args, **kwargs):
         # to the observers.
         observable = ObservableDeferred(deferred, consumeErrors=True)
         self.transactions[txn_key] = (observable, self.clock.time_msec())
+
+        # if the request fails with a Twisted failure, remove it
+        # from the transaction map. This is done to ensure that we don't
+        # cache transient errors like rate-limiting errors, etc.
+        def remove_from_map(err):
+            del self.transactions[txn_key]
```

Comment:
I'd actually do this as `self.transactions.pop(txn_key, None)` just in case the txn has already been deleted somehow. (`del` will throw if it doesn't exsit)